packages/matrix-client-server/src/account/3pid/delete.ts

```diff
@@ -34,48 +152,93 @@ const delete3pid = (clientServer: MatrixClientServer): expressAppHandler => {
+                400,
+                errMsg(
+                  'invalidParam',
+                  'Invalid medium, medium must be either email or msisdn'
+                ),
+                clientServer.logger
               )
-              const validToken = (
-                registerResponse.json() as unknown as RegisterResponseBody
-              ).token
-              const UnbindResponse = await fetch(
-                `https://${body.id_server}/_matrix/identity/v2/3pid/unbind`,
-                {
-                  method: 'POST',
-                  headers: {
-                    Authorization: `Bearer ${validToken}`,
-                    Accept: 'application/json',
-                    'Content-Type': 'application/json'
-                  },
-                  body: JSON.stringify({
-                    address: body.address,
-                    medium: body.medium
-                  })
-                }
+              return
+            }
+            if (body.medium === 'email' && !isEmailValid(body.address)) {
+              send(
+                res,
+                400,
+                errMsg('invalidParam', 'Invalid email address'),
+                clientServer.logger
               )
-              if (UnbindResponse.ok) {
-                // TODO : delete the association from the database
-                send(res, 200, {})
-              } else {
-                send(res, UnbindResponse.status, UnbindResponse.json())
-              }
+              return
+            }
+            if (body.medium === 'msisdn' && !isPhoneNumberValid(body.address)) {
+              send(
+                res,
+                400,
+                errMsg('invalidParam', 'Invalid phone number'),
+                clientServer.logger
+              )
+              return
+            }
+            let idServer: string
+            // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
+            if (typeof body.id_server === 'string' && body.id_server) {
+              idServer = body.id_server
+              deleteAndSend(
+                res,
+                body,
+                clientServer,
+                idServer,
+                data,
+                token
+              ).catch((e) => {
+                // istanbul ignore next
+                clientServer.logger.error('Error while deleting 3pid', e)
+                // istanbul ignore next
+                send(res, 500, errMsg('unknown', e), clientServer.logger)
+              })
+            } else {
+              clientServer.matrixDb
+                .get('user_threepid_id_server', ['id_server'], {
+                  user_id: data.sub,
+                  medium: body.medium,
+                  address: body.address
+                })
+                .then((rows) => {
+                  if (rows.length === 0) {
+                    clientServer.logger.error(
+                      'No id_server found corresponding to the user'
```

Comment:
Could you replace "user" by the name given by token ? It's better to set the name of user when available in  400 responses (bad request)

Reply:
Yes done